docs/function_guide.rst

```diff
@@ -0,0 +1,264 @@
+the ``test_1d_sampling.py`` regression test resembles::
+
+    {
+        1: {
+            'H_fields': ['x'],
+            'persis_info': {'rand_stream': RandomState(...) at ..., 'worker_num': 1},
+            'tag': 1,
+            'libE_info': {'H_rows': array([368])}
+        },
+
+        2: {
+            'H_fields': ['x'],
+            'persis_info': {'rand_stream': RandomState(...) at ..., 'worker_num': 2},
+            'tag': 1,
+            'libE_info': {'H_rows': array([369])}
+        },
+
+        3: {
+            'H_fields': ['x'],
+            'persis_info': {'rand_stream': RandomState(...) at ..., 'worker_num': 3},
+            'tag': 1,
+            'libE_info': {'H_rows': array([370])}
+        },
+
+        4: {
+            'H_fields': ['x'],
+            'persis_info': {'rand_stream': RandomState(...) at ..., 'worker_num': 4},
+            'tag': 1,
+            'libE_info': {'H_rows': array([371])}
+        }
+    }
+
+Based on information from the API reference above, this Work dictionary
+describes instructions for each of the four workers to call the ``sim_f``
+with data from the ``'x'`` field and a given ``'H_row'`` from the
+History array, and also pass ``persis_info``.
+
+Constructing these arrays and determining which workers are available
+for receiving data is simplified by several functions available within the
+``libensemble.tools.alloc_support`` module:
+
+.. currentmodule:: libensemble.tools.alloc_support
+.. autofunction:: avail_worker_ids
+
+Many ``alloc_f`` routines loop over the available workers returned by the above
+function to construct their Work dictionaries with the help of the following two
+functions.
+
+.. currentmodule:: libensemble.tools.alloc_support
+.. autofunction:: sim_work
+
+.. currentmodule:: libensemble.tools.alloc_support
+.. autofunction:: gen_work
+
+Note that these two functions *append* an entry in-place to the Work dictionary,
+and that additional parameters are appended to ``libE_info``.
+
+In practice, the structure of many allocation functions resemble::
+
+    Work = {}
+    ...
+    for ID in avail_worker_ids(W):
+        ...
+        if some_condition:
+            sim_work(Work, ID, chosen_H_fields, chosen_H_rows, persis_info)
+            ...
+
+        if another_condition:
+            gen_work(Work, ID, chosen_H_fields, chosen_H_rows, persis_info)
+            ...
+
+    return Work, persis_info
+
+The final three functions available in the ``alloc_support`` module
+are primarily for evaluating running generators:
+
+.. currentmodule:: libensemble.tools.alloc_support
+.. autofunction:: test_any_gen
+
+.. currentmodule:: libensemble.tools.alloc_support
+.. autofunction:: count_gens
+
+.. currentmodule:: libensemble.tools.alloc_support
+.. autofunction:: count_persis_gens
+
+Descriptions of included allocation functions can be found :doc:`here<examples/alloc_funcs>`.
+Below is the ``fast_alloc`` allocation function as an example:
```

Comment:
Maybe worth relating this to the default alloc_function - where the default (give_sim_work_first) has similar structure as fast_alloc but has some extra features that can be used in advanced cases.

Reply:
Definitely worth relating to the default alloc_f. Also note that if the alloc_f returns a 1 in the third argument, it lets the run stop. Otherwise, the alloc_f returning nothing is an error (when all workers are idle).